xla/service/gpu/gemm_algorithm_picker.cc

```diff
@@ -327,31 +330,45 @@ class GemmAutotuner {
         continue;
       }
 
+      num_algorithms_left_++; 
       if (!reference_algorithm) {
         TF_RETURN_IF_ERROR(stream_->Memcpy(&reference_buffer, OutputBuffer(),
                                            OutputBuffer().size()));
         reference_algorithm = profile_result.algorithm();
-      } else {
-        // Perform the comparison.
-        TF_ASSIGN_OR_RETURN(
-            bool outputs_match,
-            comparator.CompareEqual(stream_, /*current=*/OutputBuffer(),
+        continue;
+      } 
+      // Perform the comparison versus the reference algorithm.
+      TF_ASSIGN_OR_RETURN(
+          bool outputs_match,
+          comparator.CompareEqual(stream_, /*current=*/OutputBuffer(),
                                     /*expected=*/reference_buffer));
-        if (!outputs_match) {
-          LOG(ERROR) << "Results mismatch between different GEMM algorithms. "
-                     << "This is likely a bug/unexpected loss of precision.";
-          CHECK(!autotune_config_.should_crash_on_check_failure());
+      if (!outputs_match) {
+        LOG(ERROR) << "Results mismatch between different GEMM algorithms. "
+                   << "This is likely a bug/unexpected loss of precision.";
+        CHECK(!autotune_config_.should_crash_on_check_failure());
 
-          result.mutable_failure()->set_kind(AutotuneResult::WRONG_RESULT);
-          result.mutable_failure()->mutable_reference_gemm()->set_algorithm(
-              *reference_algorithm);
+        // By default, autotuner does NOT really skip wrong results, but 
+        // merely prints out the above error message: this may lead to a 
+        // great confusion. When should_skip_wrong_results() is set to true,
+        // solutions with accuracy problems will be disqualified.
+        auto kind = AutotuneResult::WRONG_RESULT;
+        if (autotune_config_.should_skip_wrong_results()) {
+          kind = AutotuneResult::DISQUALIFIED;
+          num_algorithms_left_--; // Decrement again since we disqualified it.
         }
+        result.mutable_failure()->set_kind(kind);
+        result.mutable_failure()->mutable_reference_gemm()->set_algorithm(
+              *reference_algorithm);
       }
     }  // for algorithms
 
     absl::StatusOr<AutotuneResult> best =
         PickBestResult(results, gemm->ToString(), hlo_module_config);
     if (best.ok()) {
+      // Return a real algorithm ID if return_algo_index is false: 
+      // e.g., in case of legacy cublas tuning.
+      if (!return_algo_index) return best; 
```

Comment:
Is this really necessary? It would be better if we always return a consistent value. Can you make this PR work with the mapping code below and remove the `return_algo_index` parameter? If not, could you please explain why?

Reply:
yes, the point is, in case of cublas-lt, we really need to get **an index** of the best solutions from the list. Later on, this algorithm_idx is used to get the actual solution [here](https://github.com/ROCm/xla/blob/01c6ba775e824415b3117a6cdf92d57518b75cc7/xla/service/gpu/runtime/gpublas_lt_matmul_thunk.cc#L145). The reason for this is because, for cublas-lt, the algorithm is essentially **an opaque object** which cannot be stored inside an HLO.

In constast, for cublas, we just take the actual algorithm since cublas' algorithm is just a 32-bit integer. Previously, we had a double conversion in algorithm picker: first from index to solution, and then from solution back to index (in case of cublas-lt). And I just simplified this path.